Adding 'Apply' action and 'Kernel Config File' setting menu to `kw upstream-patches-ui`

davidbtadokoro opened the pull request:
This PR has the objective of integrating the `kw upstream-patches-ui` feature with `kw build`, i.e., allowing the user to build a patch version of the kernel through the feature's UI. The steps involved are:

- [x] Add the 'Apply' action to apply the patch version of the kernel to a given kernel tree and a target branch
- [x] Add the 'Kernel Config File' setting to let the user define which .config file to use
- ~~Add the 'Build' action to build the patch version of the kernel to a given kernel tree and a target branch~~

It is expected that 'Apply' triggers the 'Download' action and 'Build' triggers 'Apply'.

**[UPDATE]**
This PR was reduced to not integrate the feature with `kw build`, as the feature needs to be cleaned and refactored. Integrating `kw build` before it will only increase the mess.

magalilemes commented on src/lib/lore.sh:
```diff
@@ -732,3 +735,93 @@ function save_new_lore_config()
+      warning 'Application of rejected patch was a success!'
+      warning 'Recommended examining each replaced file for unresolved conflicts and semantic.'
+    else
+      complain "Could not apply series ${series_path}"
```
I wonder if, after reaching this point,  kw should abort the `git am` operation it started, or even inform the user that they are in the middle of a `git am` and should themselves deal with this, i.e. abort, skip or solve the issue.
I do miss the error message, gone with `/dev/null 2>&1`, that would help understanding why the series could not be applied. I'm not sure, though, how this output could be presented to the user, as it doesn't seem suitable for the UI.

davidbtadokoro replied:
Thanks for the review @magalilemes and nice catch!
You are right, we should abort or inform the user that there is a `git am` in progress. I opted for the latter thinking, as the application is encapsulated in a new branch, it is good to present the user with the option to deal with it but completely forgot to warn the user about it 😞. Because of the way dialog is implemented, any output to the stdout is not displayed correctly, which is why I used `/dev/null 2>&1` to not "pollute" the dialog's screen. Have the feeling that there is a better way to do this and I'm going to look at it.

davidbtadokoro replied:
Good idea, going to work on that. I was thinking of not using
```bash
return "$ret"
```
throughout this function and returning static numbers for each error handling to make it predictable and possible to return a specific number here to signal the caller that a Yes/No screen needs to be created to prompt the user between these 3 options. What do you guys think?

davidbtadokoro commented:
@rodrigosiqueira Sorry for the delay. Working on integrating with `kw build` (hoping that it is less complicated than adding the 'Apply' action 😆). If you can, please review my new changes and look at the comment that @magalilemes made above (would appreciate your insight on the matter).

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -187,6 +189,34 @@ function show_bookmarked_series_details()
+                if [[ "$?" != 0 ]]; then
+                  create_message_box 'Error' 'Could not download patch(es):'$'\n'"- ${series['patch_title']}"$'\n'"[error message] ${output}"
+                  return
+                fi
+                output=$(apply_patchset "${series['patch_title']}" "$output" "${lore_config['kernel_tree_path']}" "${lore_config['kernel_tree_branch']}")
```
Add a newline after each `fi`.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -432,6 +462,34 @@ function show_series_details()
             fi
             ;;
+          'Apply')
+            if [[ -z "${lore_config['kernel_tree_path']}" || -z "${lore_config['kernel_tree_branch']}" ]]; then
+              create_message_box 'Error' 'The "Kernel Tree Path" and "Kernel Tree Branch" configs must be set.'$'\n''Go to Dashboard -> Settings.'
+              return
+            fi
+            message_box="The branch '${lore_config['kernel_tree_branch']}' of the Linux kernel tree located in '${lore_config['kernel_tree_path']}' "
+            message_box+='will be updated and a new branch based on it will be created for the patch application. Do you wish to proceed?'
+            create_yes_no_prompt 'Warning' "$message_box"
+            case "$?" in
+              0) # Yes
+                create_loading_screen_notification 'Applying patch(es)'$'\n'"- ${series['patch_title']}"
+                output=$(download_series "${series['patch_url']}" "${lore_config['save_patches_to']}")
+                if [[ "$?" != 0 ]]; then
+                  create_message_box 'Error' 'Could not download patch(es):'$'\n'"- ${series['patch_title']}"$'\n'"[error message] ${output}"
+                  return
+                fi
+                output=$(apply_patchset "${series['patch_title']}" "$output" "${lore_config['kernel_tree_path']}" "${lore_config['kernel_tree_branch']}")
+                if [[ "$?" != 0 ]]; then
+                  create_message_box 'Fail' 'Could not apply patch(es):'$'\n'"- ${series['patch_title']}"$'\n'$'\n'"[error message] ${output}"
+                  return
+                fi
+                create_message_box 'Success' "'${series['patch_title']}' applied."$'\n'"${output}"
+                ;;
+              1) # No
+                create_message_box '' 'Patch application aborted!'
+                ;;
+            esac
```
Is this the same code as the previous one? Can you create a common function for that?

davidbtadokoro replied:
Nice catch, this is also really bothering me as it is happening throughout the whole file. Going to work on that in a separate PR for refactoring this whole file.

rodrigosiqueira commented on etc/dialog_help/file_selection_help.txt:
```diff
@@ -0,0 +1,12 @@
+
+To move between the regions, use the <TAB> key.
+To complete the current path with the highlighted directory/file, use the <SPACE> key.
+Typing while in the Upper Box or the Lower Box alters the current path.
```
Maybe create a single commit for this string change.

rodrigosiqueira commented on src/kwlib.sh:
```diff
@@ -717,3 +717,54 @@ function get_git_repository_branches()
+  # Resetting associative array reference to prevent false configs
+  _kernel_configs=()
+
+  query_output=$(select_from 'kernel_config' '"name","description"')
```
`select_from` may fail; you must handle does cases.

rodrigosiqueira commented on src/lib/dialog_ui.sh:
```diff
@@ -373,6 +373,61 @@ function create_directory_selection_screen()
+# Returns 0 if the 'OK' button is pressed, 1 if the 'Cancel' button is pressed
+# and 2 if the 'Help' button is pressed. The string in the input box (current
+# path) is returned via the `menu_return_string` variable.
+function create_file_selection_screen()
```
This function should have its own commit.

rodrigosiqueira commented on src/lib/dialog_ui.sh:
```diff
@@ -373,6 +373,61 @@ function create_directory_selection_screen()
+# path) is returned via the `menu_return_string` variable.
+function create_file_selection_screen()
+{
+  local starting_path="$1"
+  local box_title="$2"
+  local height="$3"
+  local width="$4"
+  local flag="$5"
```
fselect does not support an extra button and label rename? If so, add those extra parameters.

rodrigosiqueira commented on src/upstream_patches_ui.sh:
```diff
@@ -338,6 +340,72 @@ function show_settings_screen()
+              if [[ "${#kernel_configs[@]}" == 0 ]]; then
+                create_message_box '' 'There are no kernel configs under kw management'
+                return
+              fi
```
newline.

davidbtadokoro commented:
@rodrigosiqueira Thanks for the comments! Going to work on them 👍

Commit:
src: lib: dialog_ui.sh: Add 'Yes/No' prompt screen

This commit adds a dialog screen that prompts the user to choose between
2 or 3 options. By default this screen has two buttons labeled 'Yes' and
'No', hence the name of the screen, and a third button can be added
passing its label as argument. Note that this function can be used for
any binary or ternary decision making.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: lib: kw_string: Add function for converting string to filename

Sometimes, we need to convert an arbitrary string into a filename.
However, there some processing may be necessary, such as:
1) Remove some bad characters
2) Replace some bad characters to change the original string the least
3) Assure that the filename isn't empty or is bigger than 255 characters

Point number 3 is based on the fact that most filesystems have a maximum
filename length of 255 characters. For reference see
https://en.wikipedia.org/wiki/Comparison_of_file_systems#Limits.

This commit adds a function to the string library for this conversion
following points 1 to 3.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -732,3 +736,180 @@ function save_new_lore_config()
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    complain "Could not switch kernel tree ${kernel_tree_path} to branch ${base_kernel_tree_branch}"
+    return "$ret"
```
Don't know all but, for example, trying `git switch` whilst rebasing you get the message
```
fatal: cannot switch branch while rebasing
```
and 128 as return.

davidbtadokoro commented on src/lib/lore.sh:
```diff
@@ -732,3 +736,180 @@ function save_new_lore_config()
+
+  new_branch_name=$(date +'%Y-%m-%d-%H-%M-%S-')
+  new_branch_name+=$(string_to_unix_filename "$series_title")
+  cmd="git -C ${kernel_tree_path} checkout -b ${new_branch_name} --quiet"
+  cmd_manager "$flag" "$cmd"
+  ret="$?"
+  if [[ "$ret" != 0 ]]; then
+    complain "Could not create new branch ${new_branch_name} based on ${base_kernel_tree_branch}"
+    return "$ret"
+  fi
```
Hmm, here I can't think of an example of creating a new branch on top of HEAD and failing. Any ideas?

Commit:
src: upstream_patches_ui: Add 'Apply' action

This commit adds the action to apply a patchset to a given Linux kernel
tree and target branch (that can be setted in the
'Dashboard'->'Settings' screen. The patchset application updates the
target branch and creates a new one based on the target branch. The new
branch name follows the pattern:
  YYYY-MM-DD-hh-mm-ss-<patchset_title_converted_to_filename>

Under the hood the patch application is done via `git am` and can fail
for any number of reasons. In case the application fails, .rej files are
generated for the files that could not be applied by `git am` and we use
`wiggle` (if installed in the system) to try apply the rejected files.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: lib: dialog_ui: Add function to create 'File Selection' screen

This commit adds a function to create a dialog screen for file
selection. Much like the directory selection screen, there are two upper
panels, the left one with the current directories and the right one with
the current files. Under the panels there is an input field where the
current path is. When hitting the 'OK' button, the path in the input
field is returned through the `menu_return_string` variable.

This commit also adds text for a help screen for the 'File Selection'
screen that can be used for with the `create_help_screen`.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

Commit:
src: upstream_patches_ui: Add 'Kernel Config File' menu

When building a Linux kernel, a fundamental file is the kernel config
file, A.K.A. .config. As we are aiming to add the possibility to build
a patchset version of the kernel through the `upstream-patches-ui`
feature, it is indispensable to have a setting to contemplate this file.

This commit adds a menu entry named 'Kernel Config File' located inside
'Dashboard'->'Settings' that allows the user to set a path to a kernel
config file.

There are two options for the user to set this through the feature:
- Choosing a user .config managed by kw (added through the
  `kernel_config_manager` feature)
- Choosing an arbitrary file from the user filesystem

The first approach has the benefit of guaranteeing that a kernel config
file is chosen and the second that the .config doesn't need to be under
kw management.

Signed-off-by: David Tadokoro <davidbtadokoro@usp.br>

rodrigosiqueira commented:
3 patches merged. Thanks
